core/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java

```diff
@@ -199,7 +199,12 @@ public void createSnapshot(final SnapshotRequest request, final CreateSnapshotLi
         final String repositoryName = request.repositoryName;
         final String snapshotName = request.snapshotName;
         validate(repositoryName, snapshotName);
-        final SnapshotId snapshotId = new SnapshotId(snapshotName); // creates a new UUID for the snapshot
+        final SnapshotId snapshotId = new SnapshotId(snapshotName, UUIDs.randomBase64UUID()); // new UUID for the snapshot
+
+        // check if the snapshot name already exists in the repository
+        if (snapshotNameExists(repositoryName, snapshotName)) {
```

Comment:
I think this check needs to go to `beginSnapshot` (where snapshot has officially become part of cluster state) to ensure that no two snapshots with same name are created in parallel. Where was this originally checked?